#! /bin/sh /usr/share/dpatch/dpatch-run
## 20_fix_read_pdo_count.dpatch by  <sascha.ittner@modusoft.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: fix read-pdo error with stoeber servo amp

@DPATCH@

diff -urNp etherlabmaster-1.5.2+hg4b0b90.orig/master/fsm_pdo_entry.c etherlabmaster-1.5.2+hg4b0b90/master/fsm_pdo_entry.c
--- etherlabmaster-1.5.2+hg4b0b90.orig/master/fsm_pdo_entry.c	2015-02-19 09:19:29.000000000 -0500
+++ etherlabmaster-1.5.2+hg4b0b90/master/fsm_pdo_entry.c	2015-05-06 10:00:04.357263437 -0400
@@ -224,7 +224,13 @@ void ec_fsm_pdo_entry_read_state_count(
         return;
     }
 
-    if (fsm->request.data_size != sizeof(uint8_t)) {
+    if (fsm->request.data_size == sizeof(uint8_t)) {
+      fsm->entry_count = EC_READ_U8(fsm->request.data);
+    } else if (fsm->request.data_size == sizeof(uint16_t)) {
+      fsm->entry_count = EC_READ_U16(fsm->request.data);
+    } else if (fsm->request.data_size == sizeof(uint32_t)) {
+      fsm->entry_count = EC_READ_U32(fsm->request.data);
+    } else {
         EC_SLAVE_ERR(fsm->slave, "Invalid data size %zu at uploading"
                 " SDO 0x%04X:%02X.\n",
                 fsm->request.data_size, fsm->request.index,
@@ -233,8 +239,6 @@ void ec_fsm_pdo_entry_read_state_count(
         return;
     }
 
-    fsm->entry_count = EC_READ_U8(fsm->request.data);
-
     EC_SLAVE_DBG(fsm->slave, 1, "%u PDO entries mapped.\n", fsm->entry_count);
 
     // read first PDO entry
